### PR TITLE
Complete receipts with data from state.db

### PR DESCRIPTION
**Pull request by pathornteng**
- [ ] I added unit tests for any code that added
- [ ] I updated the CHANGELOG.md 
- [x] All IP is original and not copied from another source
- [x] I assign all copyright to Loom Network for the code in the pull request

### CODE REVIEW

File: rpc/query_server.go [hunk 1]
```diff
@@ -1167,21 +1167,25 @@ func getReceiptByTendermintHash(
 }
 
 func completeReceipt(txResults *ctypes.ResultTx, blockResult *ctypes.ResultBlock, txReceipt *types.EvmTxReceipt) *eth.JsonTxReceipt {
+	jsonReceipt := eth.EncTxReceipt(*txReceipt)
```
Comment (enlight) [inline rpc/query_server.go]: Moving it up here means the block time won't be set in the returned receipt, doesn't it?

File: rpc/query_server.go [hunk 2]
```diff
@@ -782,8 +782,8 @@ func (s *QueryServer) EthGetTransactionReceipt(hash eth.Data) (*eth.JsonTxReceip
 	txResults, err := s.BlockStore.GetTxResult(blockResult.Block.Data.Txs[txReceipt.TransactionIndex].Hash())
 	if err != nil {
 		if strings.Contains(errors.Cause(err).Error(), "not found") {
-			// return nil response if cannot find hash
-			return nil, nil
+			// return receipts without blockresult info
```
Comment (enlight) [inline rpc/query_server.go]: If `blockStore.GetTxResult` doesn't have the data try `blockStore.GetBlockResults`.

File: rpc/query_server.go (hunk 1 above)
Comment (pathornteng) [reply rpc/query_server.go]: Move it back

File: rpc/query_server.go (hunk 2 above)
Comment (pathornteng) [reply rpc/query_server.go]: Done

File: rpc/query_server.go [hunk 3]
```diff
@@ -782,8 +782,12 @@ func (s *QueryServer) EthGetTransactionReceipt(hash eth.Data) (*eth.JsonTxReceip
 	txResults, err := s.BlockStore.GetTxResult(blockResult.Block.Data.Txs[txReceipt.TransactionIndex].Hash())
 	if err != nil {
 		if strings.Contains(errors.Cause(err).Error(), "not found") {
-			// return nil response if cannot find hash
-			return nil, nil
+			txResults, err := s.BlockStore.GetTxResultByHeightAndIndex(&height, int(txReceipt.TransactionIndex))
+			if err != nil {
+				// return receipts without blockresult info
+				return completeReceipt(nil, blockResult, &txReceipt), nil
```
Comment (enlight) [inline rpc/query_server.go]: I think we should return nil in this case, like we used to. Creating a receipt with an inaccurate status seems worse than just returning nothing.

File: store/block_store_lru_cache.go [hunk 4]
```diff
@@ -131,3 +132,24 @@ func (s *LRUBlockStoreCache) GetTxResult(txHash []byte) (*ctypes.ResultTx, error
 	}
 	return txResult, nil
 }
+
+func (s *LRUBlockStoreCache) GetTxResultByHeightAndIndex(height *int64, index int) (*ctypes.ResultTx, error) {
```
Comment (enlight) [inline store/block_store_lru_cache.go]: What's the point of this function, it doesn't check the cache, nor update it.

File: store/block_store.go [hunk 5]
```diff
@@ -150,6 +156,27 @@ func (s *TendermintBlockStore) GetTxResult(txHash []byte) (*ctypes.ResultTx, err
...
+		return nil, ErrIndexOutOfRange
+	}
+
+	return &ctypes.ResultTx{
+		Index:  uint32(index),
+		Height: *height,
+		TxResult: abci.ResponseDeliverTx{
+			Code: blockResult.Results.DeliverTx[index].Code,
+			Data: blockResult.Results.DeliverTx[index].Data,
+			Info: blockResult.Results.DeliverTx[index].Info,
```
Comment (enlight) [inline store/block_store.go]: Remember the whole `Results.DeliverTx[index] == nil` incident?

File: store/block_store.go [hunk 6]
```diff
@@ -150,6 +156,27 @@ func (s *TendermintBlockStore) GetTxResult(txHash []byte) (*ctypes.ResultTx, err
 	}, nil
 }
 
+func (s *TendermintBlockStore) GetTxResultByHeightAndIndex(height *int64, index int) (*ctypes.ResultTx, error) {
```
Comment (enlight) [inline store/block_store.go]: Adding this to the block store seems premature

File: rpc/query_server.go (hunk 3 above)
Comment (pathornteng) [reply rpc/query_server.go]: Done

File: store/block_store.go (hunk 5 above)
Comment (pathornteng) [reply store/block_store.go]: Check nil before dereferencing

File: store/block_store.go (hunk 6 above)
Comment (pathornteng) [reply store/block_store.go]: Removed

**Pull request merged**